### PR TITLE
Move relation size cache from WalIngest to DatadirTimeline

### PR DESCRIPTION
**Pull request by knizhnik**
We need relation size cache in timeline to speed up getting relation size without accessing storage.
It may be critical for creating new branch when logical size of new branch is calculated.
See discussion at https://app.slack.com/client/T026T3BRN0P/C036U0GRMRB/thread/C036U0GRMRB-1657747809.541039

### CODE REVIEW

File: pageserver/src/pgdatadir_mapping.rs [hunk 1]
```diff
@@ -154,7 +165,18 @@ impl<R: Repository> DatadirTimeline<R> {
...
+
+        // Update relation size cache
+        let mut relsize_cache = self.relsize_cache.write().unwrap();
+        if let Some((cached_lsn, _cached_nblocks)) = relsize_cache.get(&tag) {
+            if lsn >= *cached_lsn {
+                relsize_cache.insert(tag, (lsn, nblocks));
+            }
+        } else {
+            relsize_cache.insert(tag, (lsn, nblocks));
+        }
```
Comment (aome510) [inline pageserver/src/pgdatadir_mapping.rs:178]: Can be used with [`Hashmap::entry`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry) to avoid additional lookup

File: pageserver/src/pgdatadir_mapping.rs [hunk 2]
```diff
@@ -44,6 +44,8 @@ where
 
     /// Current logical size of the "datadir", at the last LSN.
     current_logical_size: AtomicIsize,
+
+    relsize_cache: RwLock<HashMap<RelTag, (Lsn, BlockNumber)>>,
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pageserver/src/pgdatadir_mapping.rs:48]: Done

File: pageserver/src/pgdatadir_mapping.rs (hunk 1 above)
Comment (knizhnik) [reply pageserver/src/pgdatadir_mapping.rs:178]: Done

**Comment by knizhnik** [comment]
> Look like `walingest::tests::test_relsize` is failing locally. Can you fix it first?

Fixed

File: pageserver/src/layered_repository.rs [hunk 3]
```diff
@@ -1128,8 +1128,6 @@ impl Timeline for LayeredTimeline {
 
     /// Look up the value with the given a key
     fn get(&self, key: Key, lsn: Lsn) -> Result<Bytes> {
-        debug_assert!(lsn <= self.get_last_record_lsn());
```
Comment (aome510) [inline pageserver/src/layered_repository.rs:1131]: hmm why do we need to remove this?

File: pageserver/src/pgdatadir_mapping.rs [hunk 4]
```diff
@@ -744,9 +786,16 @@ impl<'a, R: Repository> DatadirModification<'a, R> {
 
         self.pending_nblocks += nblocks as isize;
 
+        // Update relation size cache
+        let mut relsize_cache = self.tline.relsize_cache.write().unwrap();
+        println!(
```
Comment (aome510) [inline pageserver/src/pgdatadir_mapping.rs:791]: should be `debug!`

File: pageserver/src/pgdatadir_mapping.rs [hunk 5]
```diff
@@ -762,6 +811,14 @@ impl<'a, R: Repository> DatadirModification<'a, R> {
         let buf = nblocks.to_le_bytes();
         self.put(size_key, Value::Image(Bytes::from(buf.to_vec())));
 
+        // Update relation size cache
+        let mut relsize_cache = self.tline.relsize_cache.write().unwrap();
+        println!(
```
Comment (aome510) [inline pageserver/src/pgdatadir_mapping.rs:816]: should be `debug!`

File: pageserver/src/pgdatadir_mapping.rs [hunk 6]
```diff
@@ -781,6 +838,14 @@ impl<'a, R: Repository> DatadirModification<'a, R> {
             let buf = nblocks.to_le_bytes();
             self.put(size_key, Value::Image(Bytes::from(buf.to_vec())));
 
+            // Update relation size cache
+            let mut relsize_cache = self.tline.relsize_cache.write().unwrap();
+            println!(
```
Comment (aome510) [inline pageserver/src/pgdatadir_mapping.rs:843]: should be `debug!`

File: pageserver/src/pgdatadir_mapping.rs [hunk 7]
```diff
@@ -781,6 +838,14 @@ impl<'a, R: Repository> DatadirModification<'a, R> {
...
+                "Set cached size {} of exteded relation {} at lsn {}",
```
Comment (aome510) [inline pageserver/src/pgdatadir_mapping.rs:844]:
```suggestion
                "Set cached size {} of extended relation {} at lsn {}",
```

File: pageserver/src/layered_repository.rs (hunk 3 above)
Comment (knizhnik) [reply pageserver/src/layered_repository.rs:1131]: Because I take `lsn` from  `DatadirModification` and it can be larger than `last_record_lsn` at the moment of adding modification because `last_record_lsn` is advanced by `DatadirModification::commit`.

File: pageserver/src/pgdatadir_mapping.rs [hunk 8]
```diff
@@ -141,6 +146,18 @@ impl<R: Repository> DatadirTimeline<R> {
     pub fn get_rel_size(&self, tag: RelTag, lsn: Lsn) -> Result<BlockNumber> {
         ensure!(tag.relnode != 0, "invalid relnode");
 
+        {
+            let relsize_cache = self.relsize_cache.read().unwrap();
+            if let Some((cached_lsn, nblocks)) = relsize_cache.get(&tag) {
+                println!(
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pageserver/src/pgdatadir_mapping.rs:152]: Oh, sorry, Forgt to remove this debug printls.

File: pageserver/src/pgdatadir_mapping.rs (hunk 7 above)
Comment (knizhnik) [reply pageserver/src/pgdatadir_mapping.rs:844]: Thank you , fixed

File: pageserver/src/layered_repository.rs (hunk 3 above)
Comment (aome510) [reply pageserver/src/layered_repository.rs:1131]: R u referring to this line: https://github.com/neondatabase/neon/pull/2094/files#diff-bf5842dbfabca9cedacf26fe7e9be45370448b19b5d0191e540b1d11739f9ca1R946.

I found that previously it was
```
let last_lsn = self.timeline.get_last_record_lsn();
```

Any reasons for this change? Look like `lsn` of `DatadirModification` is also made public for this

Comment (knizhnik) [reply pageserver/src/layered_repository.rs:1131]: For the same reason: lat_record_lsn is advanced after all modifications are done. So if we use `last_record_lsn` in relation size entry, then it will be smaller than expected LSN value.

Comment (aome510) [reply pageserver/src/layered_repository.rs:1131]: Got it. In that way, https://github.com/neondatabase/neon/pull/2094/files#diff-bf5842dbfabca9cedacf26fe7e9be45370448b19b5d0191e540b1d11739f9ca1R927 should also be changed right?

Comment (knizhnik) [reply pageserver/src/layered_repository.rs:1131]: This is read only method: no reference to `DatadirModification`.
So `getlast_record_lsn` is the only what to obtain current LSN.

**Review by aome510 — APPROVED** [review]
LGTM! Thanks for all the clarifications. Look like the latest build failed in CI but probably not because of the changes in this PR. I requested a rerun. Let's delay merging to see if the tests actually pass

**Comment by knizhnik** [comment]
This PR was approved by @aome510 but at the same time yet another PR ccommitted by @aome510 #2010 conflicts with it. #2010 removes `lsn` field from `DatadirModification`. But relation size cache actually needs this lsn!
I am not sure if saving 8 bytes per modification is enough motivation for not storing lsn and making implementation of relation size cache in LayeredTimeline problematic...
So @aome510  @hlinnaka  - what do you think about restoring this field?

I just try to explain in few words why it is needed: in relation size cache we store only most recent value of relation size. Whe n relation size is requested (with some LSN) we should check if request_lsn >= cached_lsn. Otherwise cache value can not be used.

File: pageserver/src/walreceiver/walreceiver_connection.rs [hunk 9]
```diff
@@ -154,7 +154,7 @@ pub async fn handle_walreceiver_connection(
 
                 {
                     let mut decoded = DecodedWALRecord::default();
-                    let mut modification = timeline.begin_modification();
+                    let mut modification = timeline.begin_modification(endlsn);
```
Comment (aome510) [inline pageserver/src/walreceiver/walreceiver_connection.rs:157]: > I am not sure if saving 8 bytes per modification is enough motivation for not storing lsn and making implementation of relation size cache in LayeredTimeline problematic...

To provide context, it was not about saving 8 bytes, but more about reusing `DatadirModification`.  In the below code, the `modification` is reused for each walingest. Noticed that the `lsn` is different while in the loop.

In other words, this change is not enough. You should at least update `modification.lsn = lsn` in the loop.

**Comment by aome510** [comment]
> So @aome510 @hlinnaka - what do you think about restoring this field?

Related https://github.com/neondatabase/neon/pull/2172/files#diff-f290f8e7705b225e1704154be2f32b09b36ad95d82f58c5139e04f263bf5a3a6R461-R466, in which I add `last_flush_lsn` to `DatadirModification` because `get` request function may need to retrieve some data previously modified by the `DatadirModification` and `flushed` to the disk.

To be fair, I'm good with bring back the 'lsn' field.

**Comment by knizhnik** [comment]
In any cased - I rebased the patch.
@aome510  and @hlinnaka  please check if such implementation of relation cache makes sense.
It restores `lsn` field in  `DatadirModification`

**Review by aome510 — CHANGES_REQUESTED** [review]
> please check if such implementation of relation cache makes sense.
It restores `lsn` field in  `DatadirModification`

A few things needed to address with the new implementation.

First, `DatadirModification` is atomic in terms of `lsn`. So each modification needs to be stamped with a specific WAL's `lsn` before processing changes from that WAL. Changing
```rust
        let mut modification = tline.begin_modification(endpoint);
```
is **not correct** in places that reuse the modification. For example, in `import_wal`:
```rust
        let mut modification = tline.begin_modification(endpoint);
        let mut decoded = DecodedWALRecord::default();
        while last_lsn <= endpoint {
            if let Some((lsn, recdata)) = waldecoder.poll_decode()? {
                walingest.ingest_record(recdata, lsn, &mut modification, &mut decoded)?;
                last_lsn = lsn;

                nrecords += 1;

                trace!("imported record at {} (end {})", lsn, endpoint);
            }
        }
```

In other words, every WAL modification processed in the function will be stored inside the timeline as changes from the `end_lsn` WAL. 

The above code should be updated to

```diff
-        let mut modification = tline.begin_modification(endpoint);
+        let mut modification = tline.begin_modification(last_lsn);
        let mut decoded = DecodedWALRecord::default();
        while last_lsn <= endpoint {
            if let Some((lsn, recdata)) = waldecoder.poll_decode()? {
+               modification.lsn = lsn;
                walingest.ingest_record(recdata, lsn, &mut modification, &mut decoded)?;
                last_lsn = lsn;

                nrecords += 1;

                trace!("imported record at {} (end {})", lsn, endpoint);
            }
        }
```



It's quite surprising that this implementation still passed all the tests.

Second, please update the description of `begin_modification` function (https://github.com/neondatabase/neon/blob/pgdatadir_relsize_cache/pageserver/src/pgdatadir_mapping.rs#L59-L67).

Also related, I'm quite concerned with the way relsize cache is handled with `DatadirModification`. Some places use `last_record_lsn` but others use `modification.lsn` to query the relsize. IMO, it should be all `modification.lsn`, as it's possible that we'll need to query the data modified atomically by the `DatadirModification` but not yet notified to the timeline via `commit` function. Originally raised this question in https://github.com/neondatabase/neon/pull/2094#discussion_r932287997

**Comment by knizhnik** [comment]
> First, `DatadirModification` is atomic in terms of `lsn`. So each modification needs to be stamped with a specific WAL's `lsn` before processing changes from that WAL. 


Please notice that `ingest_record` updates `lsn` field:

```rust
@@ -82,6 +77,7 @@ impl<'a, T: DatadirTimeline> WalIngest<'a, T> {
         modification: &mut DatadirModification<T>,
         decoded: &mut DecodedWALRecord,
     ) -> Result<()> {
+        modification.lsn = lsn;
         decode_wal_record(recdata, decoded).context("failed decoding wal record")?;
 
```

> Second, please update the description of `begin_modification` function (https://github.com/neondatabase/neon/blob/pgdatadir_relsize_cache/pageserver/src/pgdatadir_mapping.rs#L59-L67).
> 

Will do.

> Also related, I'm quite concerned with the way relsize cache is handled with `DatadirModification`. Some places use `last_record_lsn` but others use `modification.lsn` to query the relsize. IMO, it should be all `modification.lsn`, as it's possible that we'll need to query the data modified atomically by the `DatadirModification` but not yet notified to the timeline via `commit` function. Originally raised this question in [#2094 (comment)](https://github.com/neondatabase/neon/pull/2094#discussion_r932287997)

All mutators are using e `modification.lsn`.
Readers are using `last_record_lsn` just because there no other LSN source. But it seems to be correct, isn't it?

**Comment by aome510** [comment]
> Please notice that ingest_record updates lsn field:

Ah gotcha. Thanks for the clarification.

> All mutators are using e modification.lsn.
Readers are using last_record_lsn just because there no other LSN source. But it seems to be correct, isn't it?

I don't think so, considering what will happen when we read after writing, all happen within the same modification. For example, when we call `get_relsize(rel)` right after `handle_rel_extend(rel)` with the same `rel`.

**Comment by knizhnik** [comment]
> I don't think so, considering what will happen when we read after writing, all happen within the same modification. For example, when we call get_relsize(rel) right after handle_rel_extend(rel) with the same rel. One is updated with modification.lsn but the other is queried with last_record_lsn
 
Ok, I have to agree with you. I added `lsn` parameter to `get_relsize(rel)`.

File: pageserver/src/walingest.rs [hunk 10]
```diff
@@ -64,13 +60,13 @@ impl<'a, T: DatadirTimeline> WalIngest<'a, T> {
...
             checkpoint,
             checkpoint_modified: false,
-            relsize_cache: HashMap::new(),
         })
     }
 
     ///
     /// Decode a PostgreSQL WAL record and store it in the repository, in the given timeline.
     ///
+    /// This function updates `lsn` field of `DatadirModification`
```
Comment (aome510) [inline pageserver/src/walingest.rs:69]: Look like an implementation detail, should not put in the function desc. Either put this as "//" right before the function signature or remove this line.

**Review by aome510 — APPROVED** [review]
LGTM. Thanks for all the changes

File: pageserver/src/walingest.rs (hunk 10 above)
Comment (aome510) [reply pageserver/src/walingest.rs:69]: Yeah, that makes sense

**Comment by knizhnik** [comment]
@hlinnaka , if you have no objections, I am going to merge this PR.

File: pageserver/src/pgdatadir_mapping.rs [hunk 11]
```diff
@@ -445,6 +461,18 @@ pub trait DatadirTimeline: Timeline {
...
+    fn get_cached_rel_size(&self, tag: &RelTag, lsn: Lsn) -> Option<BlockNumber>;
+
+    /// Update cached relation size if there is no more recent update
+    fn update_cached_rel_size(&self, tag: RelTag, lsn: Lsn, nblocks: BlockNumber);
+
+    /// Store cached reltion size
+    fn set_cached_rel_size(&self, tag: RelTag, lsn: Lsn, nblocks: BlockNumber);
+
+    /// Remove cached reltion size
+    fn remove_cached_rel_size(&self, tag: &RelTag);
```
Comment (hlinnaka) [inline pageserver/src/pgdatadir_mapping.rs:475]: typos:
```suggestion
    /// Get cached size of relation if it not updated after specified LSN
    fn get_cached_rel_size(&self, tag: &RelTag, lsn: Lsn) -> Option<BlockNumber>;

    /// Update cached relation size if there is no more recent update
    fn update_cached_rel_size(&self, tag: RelTag, lsn: Lsn, nblocks: BlockNumber);

    /// Store cached relation size
    fn set_cached_rel_size(&self, tag: RelTag, lsn: Lsn, nblocks: BlockNumber);

    /// Remove cached relation size
    fn remove_cached_rel_size(&self, tag: &RelTag);
```

File: pageserver/src/pgdatadir_mapping.rs [hunk 12]
```diff
@@ -457,6 +485,9 @@ pub struct DatadirModification<'a, T: DatadirTimeline> {
     /// in the state in 'tline' yet.
     pub tline: &'a T,
 
+    /// Lsn assigned by create_modification
```
Comment (hlinnaka) [inline pageserver/src/pgdatadir_mapping.rs:488]:
```suggestion
    /// Lsn assigned by begin_modification
```

**Review by hlinnaka — APPROVED** [review]
